compiler/rustc_typeck/src/check/method/mod.rs

```diff
@@ -198,6 +201,66 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let pick =
             self.lookup_probe(span, segment.ident, self_ty, call_expr, ProbeScope::TraitsInScope)?;
 
+        if span.edition() < Edition::Edition2021 {
+            if let sym::try_into = segment.ident.name {
+                if let probe::PickKind::TraitPick = pick.kind {
+                    if !matches!(self.tcx.crate_name(pick.item.def_id.krate), sym::std | sym::core)
+                    {
+                        self.tcx.struct_span_lint_hir(
+                            FUTURE_PRELUDE_COLLISION,
+                            call_expr.hir_id,
+                            call_expr.span,
+                            |lint| {
+                                let sp = call_expr.span;
+                                let trait_name =
+                                    self.tcx.def_path_str(pick.item.container.assert_trait());
+
+                                let mut lint = lint.build(&format!(
+                                    "trait method `{}` will become ambiguous in Rust 2021",
+                                    segment.ident.name
+                                ));
+
+                                if let Ok(self_expr) =
+                                    self.sess().source_map().span_to_snippet(self_expr.span)
+                                {
+                                    let derefs = "*".repeat(pick.autoderefs);
+                                    let self_adjusted = match pick.autoref_or_ptr_adjustment {
```

Comment:
it feels like this could be changed to `let autoref = mathc { "&mut", "&", .. }` and then factor out the repeated code